### PR TITLE
Allow for implicit grants and supplementary redirect_uris.

### PR DESCRIPTION
**Pull request by atrauzzi**

### CODE REVIEW

**Comment by atrauzzi** [comment]
Talked about here: https://github.com/laravel/framework/issues/8398

File: src/Two/AbstractProvider.php [hunk 1]
```diff
@@ -177,6 +177,16 @@ protected function hasInvalidState()
     {
         $session = $this->request->getSession();
 
+        if($this->request->has('code')) {
+
+            if($redirectUri = $this->request->get('redirect_uri')) {
```
Comment (GrahamCampbell) [inline src/Two/AbstractProvider.php]: cs

File: src/Two/AbstractProvider.php [hunk 2]
```diff
@@ -177,6 +177,16 @@ protected function hasInvalidState()
     {
         $session = $this->request->getSession();
 
+        if($this->request->has('code')) {
```
Comment (GrahamCampbell) [inline src/Two/AbstractProvider.php]: cs

**Comment by atrauzzi** [comment]
Sorry, CS?

**Comment by taylorotwell** [comment]
It feels weird to put this in getInvalidState(), since it has nothing to do with determining if the "state" is invalid. Could we do this differently?

**Comment by atrauzzi** [comment]
We could possibly do it by using `getInvalidState` in combination with other things that "okay" the current request.  The problem being, the current code insists that getInvalidState be...`false`?  So, we'd either be doing an `||`, or making another function like "isCurrentRequestOk".

**Comment by taylorotwell** [comment]
I think you can use the new stateless() method on the provider.

**Pull request closed without merging**